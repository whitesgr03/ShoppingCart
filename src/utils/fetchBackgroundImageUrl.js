import { initialStorage } from "../firebase-config";

const storage = initialStorage();

const fetchBackgroundImageUrl = async page => {
	const ref = storage.ref(`images/${page}/background.jpg`);
	return await ref.getDownloadURL();
};

export { fetchBackgroundImageUrl as default };
