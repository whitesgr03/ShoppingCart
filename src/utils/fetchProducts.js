import { initialFirestore } from "../firebase-config";

const firestore = initialFirestore();

const fetchProducts = async () => {
	const productRef = firestore.collection("products");
	const result = await productRef.get();

	return result.empty
		? (() => {
				throw result;
		  })()
		: result.docs.map(item => {
				const { name, price, url } = item.data();
				return {
					id: item.id,
					name,
					url,
					price,
				};
		  });
};

export { fetchProducts as default };
