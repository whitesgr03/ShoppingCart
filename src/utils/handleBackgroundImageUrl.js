import { initialStorage } from "../firebase-config";

const getBackgroundImageUrl = async page => {
	const storage = initialStorage();
	const ref = storage.ref(`images/${page}/background.jpg`);

	try {
		const url = await ref.getDownloadURL();

		return {
			message: "Get background-image url success.",
			success: true,
			data: url,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

export { getBackgroundImageUrl as default };
