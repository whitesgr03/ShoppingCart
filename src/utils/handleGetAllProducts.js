import { initialFirestore } from "../firebase-config";

const handleGetAllProducts = async () => {
	const firestore = initialFirestore();
	const productsRef = firestore.collection("products");

	const result = await productsRef.get();

	const allProductsResult = result.empty
		? []
		: result.docs.map(item => ({
				id: item.id,
				...item.data(),
		  }));

	return allProductsResult;
};

export default handleGetAllProducts;
