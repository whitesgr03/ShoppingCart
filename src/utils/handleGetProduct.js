import { initialFirestore } from "../firebase-config";

const handleGetProduct = async productId => {
	const firestore = initialFirestore();
	const productRef = firestore.collection("products").doc(productId);

	const result = await productRef.get();

	const productResult = {
		id: result.id,
		...result.data(),
	};

	return productResult;
};

export default handleGetProduct;
