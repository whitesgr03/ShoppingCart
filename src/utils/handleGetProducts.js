import { initialFirestore } from "../firebase-config";
import handlePreLoadImage from "./handlePreLoadImage";

const handleGetProducts = async () => {
	const firestore = initialFirestore();
	const productRef = firestore.collection("products");

	const result = await productRef.get();

	await Promise.all(
		result.docs.map(item => {
			const obj = { ...item.data() };
			return handlePreLoadImage(obj.url);
		})
	);

	const productsResult = result.docs.map(item => ({
		id: item.id,
		...item.data(),
	}));

	return productsResult;
};

export default handleGetProducts;
