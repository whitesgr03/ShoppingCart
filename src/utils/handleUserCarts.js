import { initialFirestore, initialAuth } from "../firebase-config";

const getUserCart = async () => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart");

	try {
		const cartItems = await cartRef.get();
		return {
			message: "Get cart success.",
			success: true,
			data: cartItems.empty
				? []
				: cartItems.docs.map(item => ({ id: item.id, ...item.data() })),
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

const addUserCartItem = async item => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;
	const { id: productId, ...product } = item;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart");

	const productRef = cartRef.doc(productId);

	try {
		await productRef.set(product);
		return {
			message: "Add to cart success.",
			success: true,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

const updateUserCartItem = async item => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart")
		.doc(item.id);

	try {
		await cartRef.update(item);
		return {
			message: "Update cart item success.",
			success: true,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

const deleteUserCartItem = async productId => {
	const firestore = initialFirestore();
	const auth = initialAuth();

	const userId = auth.currentUser.uid;

	const cartRef = firestore
		.collection("carts")
		.doc(userId)
		.collection("cart")
		.doc(productId);

	try {
		await cartRef.delete();
		return {
			message: "Delete cart item success.",
			success: true,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

export { getUserCart, addUserCartItem, updateUserCartItem, deleteUserCartItem };
