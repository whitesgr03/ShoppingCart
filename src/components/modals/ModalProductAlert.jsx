import "../../style/modals/modalProductAlert.css";
import Icon from "@mdi/react";
import { mdiLoading } from "@mdi/js";

import { useState } from "react";

import { deleteUserCartItem } from "../../utils/handleUserCart";

import PropTypes from "prop-types";

const ModalProductAlert = ({
	userId,
	product,
	behavior,
	active,
	setAppError,
	onGetUserCart,
	onOpenModule,
}) => {
	const [loading, setLoading] = useState(false);
	const handleShowCart = () => onOpenModule("cart");
	const handleRemoveItem = async () => {
		try {
			setLoading(true);
			await deleteUserCartItem(product.id, userId);
			await onGetUserCart(userId);
			handleShowCart();
		} catch (error) {
			console.error(error);
			setAppError("Service temporarily unavailable");
		} finally {
			setLoading(false);
		}
	};

	return (
		<div className={`productAlert ${active ? "active" : ""}`}>
			<div className="title" data-testid="title">
				{behavior === "add"
					? "Add product to cart"
					: "Remove product from cart"}
			</div>
			<button className="close"></button>
			<div className={"product"}>
				{product && (
					<>
						<div className="image">
							<img src={product.url} alt={product.name} />
						</div>
						<div className="description">
							<h2 className="title">{product.name}</h2>
							<p className="quantity">
								Quantity: {product.quantity}
							</p>
							<p className="price">
								Price: ${product.price.toFixed(2)}
							</p>
							<p className="subtotal">
								Subtotal (without tax): {product.quantity}
								item(s) $
								{(product.quantity * product.price).toFixed(2)}
							</p>
						</div>
					</>
				)}
			</div>
			<div className="buttons">
				{behavior === "add" ? (
					<>
						<button className="left slide" onClick={handleShowCart}>
							VIEW CART
						</button>
						<button className="right slide close">
							CONTINUE SHOPPING
						</button>
					</>
				) : (
					<>
						<button
							className="left slide"
							onClick={handleRemoveItem}
						>
							REMOVE
						</button>
						<button
							className="right slide"
							onClick={handleShowCart}
						>
							CANCEL
						</button>
					</>
				)}
			</div>
			{loading && (
				<div className="loading">
					<Icon path={mdiLoading} spin={1} size={3}></Icon>
				</div>
			)}
		</div>
	);
};

ModalProductAlert.propTypes = {
	userId: PropTypes.string,
	product: PropTypes.object,
	behavior: PropTypes.string,
	active: PropTypes.bool,
	setAppError: PropTypes.func,
	onGetUserCart: PropTypes.func,
	onOpenModule: PropTypes.func,
};

export default ModalProductAlert;
