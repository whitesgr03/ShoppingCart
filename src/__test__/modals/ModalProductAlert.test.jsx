import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { deleteUserCartItem } from "../../utils/handleUserCarts";

import {
	useModal,
	useModalDispatch,
	useCartDispatch,
} from "../../App/RootContext";

import ModalProductAlert from "../../components/modals/ModalProductAlert";

let mockNavigate = null;

jest.mock("react-router-dom", () => ({
	...jest.requireActual("react-router-dom"),
	useNavigate: jest.fn(() => mockNavigate),
}));

jest.mock("../../firebase-config");
jest.mock("../../utils/handleUserCarts");
jest.mock("../../App/RootContext");
jest.mock("../../utils/handleUserCarts");

describe("Renders Alert Component", () => {
	it("Should return ModalProductAlert DOM", async () => {
		const mockModal = {
			modal: "",
			alertProduct: {
				state: "add",
				product: {
					id: 0,
					name: "fake",
					url: "../",
					price: 19.9,
				},
			},
		};

		useModal.mockReturnValueOnce(mockModal);

		const mockIsLoad = false;
		const mockOnLoading = jest.fn();

		const { container } = render(
			<ModalProductAlert
				isLoading={mockIsLoad}
				onLoading={mockOnLoading}
			/>
		);

		expect(container).toMatchSnapshot();
	});
	it("Should show cart DOM with click event", async () => {
		const mockModal = {
			modal: "",
			alertProduct: {
				state: "add",
				product: {
					id: 0,
					name: "fake",
					url: "../",
					price: 19.9,
				},
			},
		};

		const mockModalDispatch = jest.fn();

		useModal.mockReturnValueOnce(mockModal);
		useModalDispatch.mockReturnValueOnce(mockModalDispatch);

		const mockIsLoad = false;
		const mockOnLoading = jest.fn();

		const user = userEvent.setup();

		render(
			<ModalProductAlert
				isLoading={mockIsLoad}
				onLoading={mockOnLoading}
			/>
		);

		const button = screen.getByRole("button", { name: "VIEW CART" });

		await user.pointer({ keys: "[MouseLeft]", target: button });

		const actual = screen.getByTestId("title");

		const expected = "Add product to cart";

		expect(actual).toHaveTextContent(expected);

		expect(mockModalDispatch).toBeCalledTimes(1);
	});
	it("Should successfully remove product with click event", async () => {
		const mockModal = {
			modal: "",
			alertProduct: {
				state: "remove",
				product: {
					id: 0,
					name: "fake",
					url: "../",
					price: 19.9,
				},
			},
		};

		const mockModalDispatch = jest.fn();
		const mockCartDispatch = jest.fn();

		const deleteFetchResult = {
			message: "Delete cart item success.",
			success: true,
		};

		useModal.mockReturnValueOnce(mockModal);
		useModalDispatch.mockReturnValueOnce(mockModalDispatch);
		useCartDispatch.mockReturnValueOnce(mockCartDispatch);
		deleteUserCartItem.mockReturnValueOnce(deleteFetchResult);

		const mockIsLoad = false;
		const mockOnLoading = jest.fn();

		const user = userEvent.setup();

		render(
			<ModalProductAlert
				isLoading={mockIsLoad}
				onLoading={mockOnLoading}
			/>
		);

		const button = screen.getByRole("button", { name: "REMOVE" });

		await user.pointer({ keys: "[MouseLeft]", target: button });

		const actual = screen.getByTestId("title");

		const expected = "Remove product from cart";

		expect(actual).toHaveTextContent(expected);

		expect(mockOnLoading).toBeCalledTimes(2);
		expect(mockCartDispatch).toBeCalledTimes(1);
		expect(mockModalDispatch).toBeCalledTimes(1);
	});
	it("Should failed to remove product with click event", async () => {
		mockNavigate = jest.fn();

		const mockModal = {
			modal: "",
			alertProduct: {
				state: "remove",
				product: {
					id: 0,
					name: "fake",
					url: "../",
					price: 19.9,
				},
			},
		};

		const deleteFetchResult = {
			message: "failed to remove",
			success: false,
		};

		useModal.mockReturnValueOnce(mockModal);
		deleteUserCartItem.mockReturnValueOnce(deleteFetchResult);

		const mockIsLoad = false;
		const mockOnLoading = jest.fn();

		const user = userEvent.setup();

		render(
			<ModalProductAlert
				isLoading={mockIsLoad}
				onLoading={mockOnLoading}
			/>
		);

		const button = screen.getByRole("button", { name: "REMOVE" });

		await user.pointer({ keys: "[MouseLeft]", target: button });

		const actual = screen.getByTestId("title");

		const expected = "Remove product from cart";

		expect(actual).toHaveTextContent(expected);

		expect(mockNavigate).toBeCalledTimes(1);
		expect(mockOnLoading).toBeCalledTimes(2);
	});
});
