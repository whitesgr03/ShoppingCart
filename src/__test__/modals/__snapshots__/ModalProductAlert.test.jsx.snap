// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Alert Component Should return ModalProductAlert DOM 1`] = `
<div>
  <div
    class="productAlert"
  >
    <div
      class="title"
      data-testid="title"
    >
      Add product to cart
    </div>
    <button
      class="close"
    />
    <div
      class="product"
    >
      <div
        class="image"
        style="height: 0px;"
      >
        <img
          alt="fake"
          src="../"
        />
      </div>
      <div
        class="description"
      >
        <h2
          class="title"
        >
          fake
        </h2>
        <p
          class="quantity"
        >
          Quantity: 
        </p>
        <p
          class="price"
        >
          Price: $
          19.90
        </p>
        <p
          class="subtotal"
        >
          Subtotal (without tax): 
          item(s) $
          NaN
        </p>
      </div>
    </div>
    <div
      class="buttons"
    >
      <button
        class="left slide"
      >
        VIEW CART
      </button>
      <button
        class="right slide close"
      >
        CONTINUE SHOPPING
      </button>
    </div>
  </div>
</div>
`;
